File: backend/app/models/opportunity.py
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text
)
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship
)

from app.db.base import Base


class Opportunity(Base):
    __tablename__ = "opportunities"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True
    )

    organization_id: Mapped[int] = mapped_column(
        ForeignKey("organization_profiles.user_id"),
        nullable=False
    )

    title: Mapped[str] = mapped_column(
        String(255),
        nullable=False
    )

    description: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )

    opportunity_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False
    )

    location: Mapped[str] = mapped_column(
        String(255),
        nullable=False
    )

    skills_required: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )

    application_deadline: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow
    )

    organization = relationship(
        "OrganizationProfile",
        back_populates="opportunities"
    )

File: frontend/src/lib/validations/opportunity.ts
import { z } from "zod";

export const opportunitySchema =
  z.object({
    title: z
      .string()
      .min(
        1,
        "Title is required"
      ),

    description: z
      .string()
      .min(
        1,
        "Description is required"
      ),

    opportunity_type:
      z.string().min(
        1,
        "Opportunity type is required"
      ),

    location: z
      .string()
      .min(
        1,
        "Location is required"
      ),

    skills_required:
      z.string().min(
        1,
        "Skills are required"
      ),

    application_deadline:
      z.string().min(
        1,
        "Deadline is required"
      ),
  });

export type OpportunityFormData =
  z.infer<
    typeof opportunitySchema
  >;